File: Python/backjoon/5639.py
import sys, os

sys.setrecursionlimit(10*8)

class Node:
    def __init__(self, value: int) -> None:
        self.value = value
        self.left = None
        self.right = None

def insert(head: Node, node: Node):
    if head == None:
        return node

    if node.value < head.value:
        head.left = insert(head.left, node)

    else:
        head.right = insert(head.right, node)

    return head

root = None

while True:
    try:
        num = int(sys.stdin.readline())
        node = Node(num)

        root = insert(root, node)
    except:
        break

def recursive(head: Node):
    if head.left != None:
        recursive(head.left)

    if head.right != None:
        recursive(head.right)

    print(head.value)

recursive(root)

File: CPP/Backjoon/5639.cpp
#include <iostream>

struct Node {
    Node *left, *right;
    int value;
    Node(int value) {
        this->value = value;
        this->left = nullptr;
        this->right = nullptr;
    }
};

Node* insert(Node* head, Node* node) {
    if (head == nullptr) return node;

    if (node->value < head->value)
        head->left = insert(head->left, node);
    else
        head->right = insert(head->right, node);

    return head;
}

void DFS(Node* node) {
    if (node == nullptr) return;

    DFS(node->left);
    DFS(node->right);

    std::cout << node->value << "\n";
}

int main() {
    Node* head = nullptr;
    int value;
    while (scanf("%d", &value) != EOF) {
        Node* node = new Node(value);
        head = insert(head, node);
    }

    DFS(head);

    return 0;
}